### DISCUSSION TITLE
thread/pianxiangsuo

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#285] [Announcements]
# thread/pianxiangsuo

偏向锁是 Java 虚拟机（JVM）中一种用于提高程序运行效率的锁机制。在许多应用场景中，大部分锁是只被一个线程所访问的。如果锁对象始终被一个线程访问，那么同步操作显得非常多余。为了减少这样的重量级操作，Java 6 引入了偏向锁的概念。

https://javabetter.cn/thread/pianxiangsuo.html

**Comment by Randomyyds** [comment]
这个偏向锁还是没看懂，那个场景三那里，主线程不是都退出同步代码块了吗，为啥还是为升级成轻量锁，这与后面说的
"线程不存活，或者活着的线程退出了同步块，很简单，直接撤销偏向就好了"以及那个流程图中表述的也都不一样啊

**Reply by Kalfredwv** [reply]
+1 而且感觉关于偏向锁到轻量锁的升级情形 和 前一章说的”虽然有竞争但偏向锁不升级只是重新修改markword指向的线程ID“也不太一样

**Reply by Liucc-123** [reply]
因为主线程进入同步代码块后，对象 MarkWord 存入主线程的线程 id，当主线程退出、子线程进入同步块后，发现对象锁 MarkWord 存放的不是自己的线程 id。此时，就发生了锁的争抢，从偏向锁升级为轻量级锁。

**Reply by LFAW233** [reply]
一般的已偏向的偏向锁是不会重偏向的。线程T2在访问偏向T1的偏向锁时，会直接被阻塞，等待检查点，检查点中执行revoke_bias，撤销偏向锁。一般撤销完毕是无锁状态，检查点结束后T2恢复运行，便会进入锁，使无锁变为轻量级锁。还有种情况，T2被阻塞时并未修改偏向锁的markword，因此T1仍然能够进入偏向锁，如果检查点时T1仍在锁中，那么撤销完偏向锁会直接升级为轻量级锁，检查点结束后T2开始运行，如果T1还没有出来，那么T2就会开始自旋。

一般重偏向的情况只会发生在BulkRebias后，未被更新epoch的线程上。如果要对这些线程撤销偏向或者重偏向，那么不需要等待检查点，只需要直接对其进行CAS就行。

**Comment by LFAW233** [comment]
JDK8源码都看了一遍了，你告诉我偏向锁默认被弃用了？？我真昏过去。。。

**Comment by lumole** [comment]
为啥这文章这么晦涩难懂啊

**Reply by DavidYin66** [reply]
太失望了，这个写的根本就不想让你看懂，已经决定换教程来学了
